### PR TITLE
Implement backend configuration sorting based on dependencies

### PR DESCRIPTION
**Pull request by ksysoev**
# Decription:
This pull request adds functionality to sort the backend configuration based on dependencies between calls. It introduces a new function sortBackends that takes a slice of BackendConfig as input and returns a sorted slice of BackendConfig. The sorting is done by checking the DependsOn field of each BackendConfig and ensuring that there are no circular dependencies. If a circular dependency is detected, an error is returned. This sorting logic is then applied in the NewCallsRepository function to sort the BackendConfig slice before processing the requests.

## Notable Changes:

- Added dependency handling to the Composer struct and functions in pkg/core/composer/composer.go.
- Updated test cases to reflect changes in the Composer struct in pkg/core/composer/composer_test.go.
- Modified interfaces to handle dependencies in pkg/core/handler/handler.go.
- Updated Handler methods to incorporate dependency-based request processing.
- Refactored request generation and handling logic.

### CODE REVIEW

**Comment by ksysoev** [comment]
@shan-96 If you have time could you help to review the PR

File: pkg/repo/callrepo.go [hunk 1]
```diff
@@ -79,3 +87,71 @@ func NewCallsRepository(cfg *CallsConfig) (*CallsRepository, error) {
...
+		}
+
+		visited[v] = true
+		recStack[v] = true
+
+		for _, u := range graph[v] {
+			if err := dfs(u); err != nil {
+				return err
+			}
+		}
```
Comment (shan-96) [inline pkg/repo/callrepo.go:119]: Here are my suggestions to this function

1. use map for quick lookup
2. some more error messages
3. use defer for cleanup

```suggestion
// topSortDFS performs a topological sort on a slice of BackendConfig using Depth-First Search (DFS).
// It takes a slice of BackendConfig as input and returns a sorted slice of BackendConfig and an error.
// It returns an error if a circular dependency is detected among the BackendConfig elements.
// Each BackendConfig element must have a unique ResponseBody and a list of dependencies specified in DependsOn.
func topSortDFS(be []BackendConfig) ([]BackendConfig, error) {
	graph := createDepGraph(be)
	visited := make(map[string]bool)
	recStack := make(map[string]bool)
	sorted := []BackendConfig{}

	// Create a map for quick index lookup
	indexMap := make(map[string]int)
	for i, b := range be {
		indexMap[b.ResponseBody] = i
	}

	var dfs func(string) error
	dfs = func(v string) error {
		if recStack[v] {
			return fmt.Errorf("circular dependency detected at %s", v)
		}
		if visited[v] {
			return nil
		}

		visited[v] = true
		recStack[v] = true
		defer func() { recStack[v] = false }() // Ensure recStack is reset

		for _, u := range graph[v] {
			if err := dfs(u); err != nil {
				return err
			}
		}

		sorted = append(sorted, be[indexMap[v]])
		return nil
	}

	for k := range graph {
		if err := dfs(k); err != nil {
			return nil, err
		}
	}

	return sorted, nil
}
```

File: runtime/config.yaml [hunk 2]
```diff
@@ -30,4 +30,28 @@ api:
               "website_status": 1,
               "req_id": {{ .ReqID }}
             }
-      
\ No newline at end of file
+    - method: "chain"
```
Comment (shan-96) [inline runtime/config.yaml:33]: maybe I am missing something but how does the backend understand what dependency handling to use?
i.e. Aggregate or Chaining?

**Comment by shan-96** [comment]
Added a question and a comment. Rest LGTM

File: runtime/config.yaml (hunk 2 above)
Comment (ksysoev) [reply runtime/config.yaml:33]: @shan-96 method here stands for declaration of api call, basically if any of sub calls has declarations dependencies, then in this case they will be call with respect of dependencies. 

you can test it with these calls 

```
{"method":"chain", "req_id": 123, "passthrough":{"key":1}}
```

```
{"method":"aggregate", "params": {"country": "id"}, "req_id": 123, "passthrough":{"key":1}}
```

**Pull request merged**